recidiviz/ingest/us_ny/us_ny_scraper.py

```
@@ -41,1226 +37,276 @@
-            parole_elig_date=record.parole_elig_date,
-            parole_hearing_date=record.parole_hearing_date,
-            parole_hearing_type=record.parole_hearing_type,
-            race=record.race,
-            region=record.region,
-            sex=record.sex,
-            surname=record.surname,
-            given_names=record.given_names)
-
-        return snapshot
+        data_extractor = DataExtractor(self.mapping_filepath)
+        ingest_info = data_extractor.extract_and_populate_data(content,
+                                                               ingest_info)
+
+        # Get release date, if released
+        if ingest_info.person[0].booking[0].custody_status == 'RELEASED':
+            release_date_string = content.xpath(
+                '//*[@headers="t1l"]')[0].text.split()[0]
+            ingest_info.person[0].booking[0].release_date = release_date_string
+
+        # Parse charge information
+        for row in content.xpath('//*[@id="ii"]/table[2]/tr')[1:]:
+            charge_name = row.xpath('td[1]')[0].text.strip().rstrip()
+
+            if not charge_name:
+                break
+
+            # Change 'class' into 'level'
+            charge_class_str = row.xpath('td[2]')[0].text
+            charge_level = None
+            if charge_class_str.startswith('A'):
+                charge_level = 'ONE'
+            elif charge_class_str.startswith('B'):
+                charge_level = 'TWO'
+            elif charge_class_str.startswith('C'):
+                charge_level = 'THREE'
+            elif charge_class_str.startswith('D'):
+                charge_level = 'FOUR'
+            elif charge_class_str.startswith('E'):
+                charge_level = 'FIVE'
+
+            # Get the degree
+            charge_degree = charge_name.split()[-1]
+            try:
+                _ = ChargeDegree.from_str(charge_degree)
+                charge_name = ' '.join(charge_name.split()[:-1])
+            except KeyError:
```

How exceptional is this? I don't love use exceptions for non-exceptional control flow if it can be avoided.

---

This is, as of now, the @zacharylawrence approved method for checking whether a value is in the map and will be mapped properly. We talked about adding a 'check_valid' method to constant classes if this pattern has to repeat often.

---

Yeah, this is a good point @jessex. What alternatives would you prefer? Should `ChargeDegree.from_str()` just return `None` in that case? Forcing users to specifically check if a valid value is returned?

I think it's reasonable consider a failed `from_str()` to throw an exception but we should probably be using our own exception type instead of KeyError?

One case where `from_str()` failing with an exception is nice is during the data_conversion step where failure to parse a string into an enum means that exception should bubble up and prevent us from writing to the database.

---

I think this last point is important. I like the idea of throwing our own exception to bubble up instead of writing to the DB but silently nulling fields that don't conform. Filed #361.